File: frontend/js/my_timetable.js
const API_BASE_URL = 'http://localhost:5000/api';

const DAY_LABELS = {
    Monday: '月曜日',
    Tuesday: '火曜日',
    Wednesday: '水曜日',
    Thursday: '木曜日',
    Friday: '金曜日',
};

function escapeHtml(str) {
    if (!str) return '';
    const div = document.createElement('div');
    div.textContent = str;
    return div.innerHTML;
}

async function fetchAPI(endpoint, options = {}) {
    try {
        const response = await fetch(`${API_BASE_URL}${endpoint}`, {
            headers: { 'Content-Type': 'application/json' },
            credentials: 'include',
            ...options,
        });
        const data = await response.json().catch(() => ({}));
        if (!response.ok) {
            const err = new Error(data.error || `HTTP ${response.status}`);
            err.status = response.status;
            err.payload = data;
            throw err;
        }
        return data;
    } catch (error) {
        console.error('API Error:', error);
        throw error;
    }
}

async function init() {
    const loading = document.getElementById('loading');
    const table = document.getElementById('my-timetable-table');
    const noData = document.getElementById('no-data');
    const errorEl = document.getElementById('error');
    const userInfo = document.getElementById('user-info');
    const tbody = document.getElementById('my-timetable-list');

    loading.style.display = 'block';
    table.style.display = 'none';
    noData.style.display = 'none';
    errorEl.style.display = 'none';

    try {
        // ログイン状態確認
        const user = await fetchAPI('/auth/me');
        if (user.role !== 'teacher') {
            userInfo.textContent = `${escapeHtml(user.username)} (${user.role})`;
            loading.style.display = 'none';
            errorEl.innerHTML = `この画面は教員専用です。<a href="admin_dashboard.html" style="color: #2563eb; text-decoration: underline;">管理者ダッシュボード</a> をご利用ください。`;
            errorEl.style.display = 'block';
            return;
        }

        const teacherName = user.teacher_name || user.username;
        userInfo.textContent = `${escapeHtml(teacherName)} 先生`;

        // 自分の時間割取得
        const timetables = await fetchAPI('/my/timetable');
        loading.style.display = 'none';

        if (!timetables || timetables.length === 0) {
            noData.style.display = 'block';
            return;
        }

        tbody.innerHTML = timetables.map(t => `
            <tr>
                <td>${DAY_LABELS[t.day_of_week] || escapeHtml(t.day_of_week)}</td>
                <td><strong>${escapeHtml(String(t.period))}限</strong></td>
                <td><strong>${escapeHtml(t.subject || t.subject_name || '科目ID:' + t.subject_id)}</strong></td>
                <td>${escapeHtml(t.classroom || t.classroom_name || '教室ID:' + t.classroom_id)}</td>
            </tr>
        `).join('');

        table.style.display = '';
    } catch (err) {
        loading.style.display = 'none';
        if (err.status === 401) {
            window.location.href = '../index.html';
            return;
        }
        errorEl.textContent = '時間割の取得に失敗しました: ' + err.message;
        errorEl.style.display = 'block';
    }
}

async function handleLogout() {
    try {
        await fetchAPI('/auth/logout', { method: 'POST' });
    } catch (e) {
        console.error('Logout error:', e);
    }
    window.location.href = '../index.html';
}

document.addEventListener('DOMContentLoaded', init);

File: backend/routes/my_timetable.py
from flask import Blueprint, jsonify, session
from database import db
from models.timetable import Timetable
from models.user import User
from utils.auth import current_user

my_timetable_bp = Blueprint('my_timetable_bp', __name__)

DAY_ORDER = {
    'Monday': 1,
    'Tuesday': 2,
    'Wednesday': 3,
    'Thursday': 4,
    'Friday': 5,
}


@my_timetable_bp.route('', methods=['GET'])
@my_timetable_bp.route('/', methods=['GET'])
def get_my_timetable():
    """教員本人の時間割を取得するエンドポイント"""
    user = current_user()
    if not user or not user.is_active:
        return jsonify({'error': 'unauthenticated'}), 401

    if user.role != 'teacher':
        return jsonify({'error': 'forbidden'}), 403

    if not user.teacher_id:
        return jsonify({'error': 'teacher_profile_not_linked'}), 400

    # サーバー側の user.teacher_id のみを信頼し、クエリパラメータは使用しない
    timetables = Timetable.query.filter_by(teacher_id=user.teacher_id).all()

    # 曜日・時限順にソート
    timetables.sort(key=lambda t: (DAY_ORDER.get(t.day_of_week, 99), t.period))

    result = []
    for t in timetables:
        subj_name = t.subject.name if t.subject else None
        room_name = t.classroom.name if t.classroom else None
        result.append({
            'id': t.id,
            'teacher_id': t.teacher_id,
            'day_of_week': t.day_of_week,
            'period': t.period,
            'subject_id': t.subject_id,
            'subject': subj_name,
            'subject_name': subj_name,
            'grade': t.subject.grade if t.subject else None,
            'classroom_id': t.classroom_id,
            'classroom': room_name,
            'classroom_name': room_name,
            'is_online': t.is_online,
        })

    return jsonify(result), 200
